state/traverse.go

```diff
@@ -7,27 +7,59 @@ import (
 	"github.com/onflow/flow-go/storage"
 )
 
+// functor that will be get called on each block header when traversing blocks.
+type onVisitBlock = func(header *flow.Header) (bool, error)
+
 // Traverse traverses a chain segment beginning with the start block (inclusive)
-// and ending with the end block (inclusive). Blocks are traversed in reverse
+// Blocks are traversed in reverse
 // height order, meaning the end block must be an ancestor of the start block.
 // The callback is called for each block in this segment.
-func Traverse(headers storage.Headers, start, end flow.Identifier, callback func(header *flow.Header) error) error {
-
-	nextID := start
+// Return value of callback is used to decide if it should continue or not.
+func Traverse(headers storage.Headers, startBlockID flow.Identifier, visitor onVisitBlock) error {
+	ancestorID := startBlockID
 	for {
-		// retrieve the next block in the segment and pass it to the callback
-		next, err := headers.ByBlockID(nextID)
+		ancestor, err := headers.ByBlockID(ancestorID)
 		if err != nil {
-			return fmt.Errorf("could not get segment block (id=%x): %w", nextID, err)
+			return fmt.Errorf("could not get ancestor header (%x): %w", ancestorID, err)
+		}
+
+		shouldContinue, err := visitor(ancestor)
+		if !shouldContinue {
+			break
 		}
-		err = callback(next)
 		if err != nil {
-			return fmt.Errorf("error in callback: %w", err)
+			return err
 		}
+		ancestorID = ancestor.ParentID
+	}
+	return nil
+}
 
-		if nextID == end {
-			return nil
-		}
-		nextID = next.ParentID
+// TraverseParentFirst traverses a chain segment beginning with the last block (non inclusive)
+// and ending with start block
+// Implements a recursive descend to last block identified by `stopBlockID` and then calling
+// visitor callback for every block way back up to `startBlockID`
+func TraverseParentFirst(headers storage.Headers, startBlockID, stopBlockID flow.Identifier,
+	visitor func(header *flow.Header) error) error {
+	ancestor, err := headers.ByBlockID(startBlockID)
+	if err != nil {
+		return fmt.Errorf("could not get ancestor header (%x): %w", startBlockID, err)
+	}
+
+	if ancestor.ID() == stopBlockID {
+		return nil
+	}
```

Comment:
```suggestion

	if startBlockID == stopBlockID {
		return nil
	}

	ancestor, err := headers.ByBlockID(startBlockID)
	if err != nil {
		return fmt.Errorf("could not get ancestor header (%x): %w", startBlockID, err)
	}
```

save an ID computation